### PR TITLE
Fix setuptools>=64 import hooks

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

This needs a test. Refs: https://github.com/PyCQA/pylint/issues/7306 and https://github.com/pypa/setuptools/issues/3518.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

### CODE REVIEW

**Comment by DanielNoord** [comment]
@jacobtylerwalls This fixes the test that started failing after https://github.com/PyCQA/astroid/pull/1714

Would you mind taking a look at the PR and the linked issues/discussions? I'm hesitant to go forward, but this actually seems like it fixes some issues.

**Comment by jacobtylerwalls** [comment]
I'd like to provide input, but the fall is going to be very busy for me. I'm not sure when I'll have the time to take a look. I'm very hesitant on doing anything additional touching the import system for 2.15.

**Comment by DanielNoord** [comment]
> I'd like to provide input, but the fall is going to be very busy for me. I'm not sure when I'll have the time to take a look. I'm very hesitant on doing anything additional touching the import system for 2.15.

Let's definitely not put this in `2.15`. However, since this is going to affect myself as well when I update my local setuptools installation at some point I'd like to get this on `astroid` `main` at least. I'll try and see if I can get some more context in the meantime.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Would you be okay with putting this in the next release? This will be necessary for `pylint` not to be incompatible with `setuptools>=64` and I'd like to get feedback on it from users before everybody is on `>=64`.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Sure but I think this need a rebase the changelog is not in 2.14.0

**Comment by DanielNoord** [comment]
> Sure but I think this need a rebase the changelog is not in 2.14.0

Yeah I will prepare the PR for final review then. Thanks!

**Comment by Pierre-Sassoulas** [comment]
On closer inspection (seeing Jacob's comment) this feel like something a little risky if we want to release 2.16.0 soon and not do another beta release.  Maybe we should do an rc release ?

**Comment by DanielNoord** [comment]
> On closer inspection (seeing Jacob's comment) this feel like something a little risky if we want to release 2.16.0 soon and not do another beta release. Maybe we should do an rc release ?

Yeah I don't know. Based on the tests I think this should be fine, but we have been wrong before. If anything we can just quickly revert the change and release a patch `astroid` release? There is no code that builds upon this (both in `astroid` or `pylint`) so I don't foresee many merge issues when back porting.

**Comment by Pierre-Sassoulas** [comment]
I'm afraid 😄 Let's wait for a review by Jacob, forcing a namespace change last minute can ruin the effort we made with two beta releases and make our lives difficult (when we lived without this for months).

**Comment by jacobtylerwalls** [comment]
Sorry for not prioritizing this yet. @DanielNoord would you be able to write a (short) summary of why you believe this is the correct approach? That would help me get into this. (There has been more discussion on the linked PyPA issue.)

**Comment by DanielNoord** [comment]
The idea is that we actually have a quite an advanced import system compared to other tools and are able to follow imports (somewhat successfully).
However, when installing editable installs with `pip` a new type of `_SPEC_FINDERS` is returned. We don't have support for those but we can quite easily add them by telling our import system that we should handle them as normal (namespace) packages.
I chose to use the `__name__` attribute of the finders as we then don't build in dependencies on `pip` and `setuptools` etc. but can just infer the type of package from the finder's name.

The test case that now succeeds is notable as it shows that with this PR we actually are able to infer more imports than we previously could.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [inline astroid/interpreter/_import/spec.py:427]: No tests fail when I revert this change.

Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:427]: This is because of the typing change of `_find_spec_with_path`. `MetaPathFinder` is not a subclass of `Finder`, see:
https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [inline astroid/interpreter/_import/spec.py:386]: Should we set location to `""` or `None` rather than `spec.origin`?

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [inline astroid/interpreter/_import/spec.py:372]: I think we have to guard against `find_spec` being absent. It was only recently added to `six._SixMetaPathImporter`, see https://github.com/benjaminp/six/pull/352.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:427]: I'm still confused: won't this cause a behavior change across versions? On Python 3.9 this condition will match; on Py3.10 it won't.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [inline astroid/interpreter/_import/spec.py:360]: I'm not so sure we can limit the type of the return object now. The `six._SixMetaPathImporter` inherits directly from `object`:

```py
(Pdb) meta_finder.__class__.mro()
[<class 'urllib3.packages.six._SixMetaPathImporter'>, <class 'object'>]
```

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Thanks for the description, that really helped focus my review. This is looking really good. One requested change on guarding against a missing `find_spec` and then some questions.

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:372]: Hm, this is an issue within `six` then..

https://github.com/python/typeshed/blob/b9270da7b083f08a86e4366b3cf284e627a63873/stdlib/sys.pyi#L45

The `stdlib` is typed with a protocol to signal that this should indeed inherit from the `abc`. I'll make us use a `Protocol` as well.

Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:372]: I think part of the issue is that PEP 302 doesn't mention `find_spec`, so there could be code in the wild without it.

**Comment by jacobtylerwalls** [comment]
Also, I noticed two failures in pylint with this change, including one crash and one `astroid-error`:

```py
pytest -k private-import

Exception on node <ImportFrom l.33 at 0x1144f4790> in file '/Users/jacobwalls/pylint/tests/functional/ext/private_import/private_import.py'
Traceback (most recent call last):
  File "/Users/.../pylint/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/Users/.../pylint/pylint/checkers/imports.py", line 562, in visit_importfrom
    self._add_imported_module(node, f"{imported_module.name}.{name}")
  File "/Users/.../pylint/pylint/checkers/imports.py", line 894, in _add_imported_module
    elif not astroid.modutils.is_standard_module(importedmodname):
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/modutils.py", line 528, in is_standard_module
    filename = file_from_modpath([modname])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/modutils.py", line 334, in file_from_modpath
    return file_info_from_modpath(modpath, path, context_file).location
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/modutils.py", line 385, in file_info_from_modpath
    return _spec_from_modpath(modpath, path, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/modutils.py", line 595, in _spec_from_modpath
    found_spec = spec.find_spec(modpath, path)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/interpreter/_import/spec.py", line 423, in find_spec
    finder, spec = _find_spec_with_path(
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/.../astroid/astroid/interpreter/_import/spec.py", line 372, in _find_spec_with_path
    spec = meta_finder.find_spec(modname, submodule_path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/_pytest/assertion/rewrite.py", line 92, in find_spec
    if self._early_rewrite_bailout(name, state):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/_pytest/assertion/rewrite.py", line 193, in _early_rewrite_bailout
    path = PurePath(*parts).with_suffix(".py")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 694, in with_suffix
    raise ValueError("%r has an empty name" % (self,))
ValueError: PurePosixPath('.') has an empty name

pytest -k non_init_parent_called

E       AssertionError: Wrong results for file "non_init_parent_called":
E       
E       Expected in testdata:
E          6: import-error
E         14: non-parent-init-called
E         22: no-member
E         27: no-member
E         50: no-member
E       
E       Unexpected in testdata:
E          1: astroid-error
```

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:372]: See my last commit, would that be an okay solution?

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:400]: Why? Isn't that "sort of" right?

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:441]: Yeah it might, it is just really hard to make this type safe..
`_MetaPathFinder` doesn't have this method so we can't safely call it.

We could use `hasattr` but that also feels unsatisfactory..

**Comment by DanielNoord** [comment]
> ```python
> PurePosixPath
> ```

Feels like we should expand "." to the current working directory? 😓
That sounds like more issues..

We can switch that logic around? Only use the `meta_finders` in `sys.meta_path` that we know we support?

**Comment by jacobtylerwalls** [comment]
Interesting that the failure is coming from `pytest` itself.

**Comment by DanielNoord** [comment]
> Interesting that the failure is coming from `pytest` itself.

They probably have their own `MetaPathFinder` to handle all the `fixture` magic which we then try to use to locate an import.
That's why I'm debating on moving the check for known `MetaPathFinders` to above `line 372, in _find_spec_with_path`. That should resolve this I think.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:400]: I think it's supposed to be `None` for namespace packages and a string for others:
https://github.com/jacobtylerwalls/astroid/commit/3ff65be6c9e3a4480cd9338a68f418cb578e347b

So we could switch on the `module_type` and change to `None` -> `""` if PY_SOURCE. But I'm interested in your opinion.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:441]: Ah. I think we can't let type-checking considerations cause behavior differences.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:386]: I guess I still don't understand, an importer without a find_spec method, such as [older vendored versions of six](https://sourcegraph.com/github.com/cronyo/cronyo/-/blob/cronyo/vendor/urllib3/packages/six.py?subtree=true) will still raise an uncaught AttributeError on `spec = meta_finder.find_spec(modname, submodule_path)`.

**Comment by jacobtylerwalls** [comment]
> That's why I'm debating on moving the check for known MetaPathFinders to above line 372, in _find_spec_with_path. That should resolve this I think.

I'd be curious to see the pylint test results: I worry about giving new priority to the builtin importers, if that's what you're suggesting:

```py
>>> sys.meta_path
[<_distutils_hack.DistutilsMetaFinder object at 0x104452070>, <class '_frozen_importlib.BuiltinImporter'>, <class '_frozen_importlib.FrozenImporter'>, <class '_frozen_importlib_external.PathFinder'>]
```

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:386]: The issue is that `find_module` has been deprecated since 3.4:
https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_module

Anyway, I'll keep the typing and use a `hasattr`. That seems the most "safe" while also correct.

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:400]: It feels like this mimics `has_location`, see https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec.has_location.

Looking for `.location` it seems like only (frozen) builtins don't have their `location` set? And we use it in `ast_from_module_name` so we need it to be set to whatever `origin` is?

File: astroid/interpreter/_import/spec.py
Comment (DanielNoord) [reply astroid/interpreter/_import/spec.py:441]: I'm dumb. `Finder != Finder`. The `Finder` in the `isinstance` check is actually our own implementation of `Finder`, which has `contribute_to_path`. `MetaPathFinders` don't so we can't call the method here.

Previously all `Finders` were of our own type but with this change this is no longer the case so we need to explicitly check this.

**Comment by DanielNoord** [comment]
@jacobtylerwalls See my last change, this is what I meant with changing the order of checks. We still rely on our own 4 `Finders` initially, but if that doesn't work we start trying `MetaPathFinders` for which we now stuff works.

This means that we would need to add new support by hand, but it is much less error prone.

`pylint` test suite passed with https://github.com/PyCQA/astroid/pull/1752/commits/f889b5d6bb49278d9736b799172f19b3f09631df.

**Comment by DanielNoord** [comment]
If we agree that this is the way to go I'll add support for `SixMetaPathImporter` so the test case that I "unfailed" in this PR will pass again.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [reply astroid/interpreter/_import/spec.py:408]: This is moot now that we're not setting `PY_SOURCE`. I was just wondering about the 0.05% of cases where we hit `KeyError` and assumed `PY_SOURCE` and then still set location to `None` instead of `""`, but it looks like after latest commits we aren't doing setting `PY_SOURCE`.

**Review by jacobtylerwalls — APPROVED** [review]
Looks good, thanks! (A test would be nice.)

**Comment by DanielNoord** [comment]
> Looks good, thanks! (A test would be nice.)

I think the `six` test serves as a good test for the general logic. We can now import and find that module where we previously couldn't as we support the `_SixMetaPathImporter` now.
As for `setuptools` I don't think there is a good way to test it without introducing a dependency on it, which I don't really want.

I have tested this with https://github.com/jooola/pylint-import-error-bug to confirm that this branch does indeed fix the issue mentioned in the linked issue.

File: astroid/interpreter/_import/spec.py
Comment (jacobtylerwalls) [inline astroid/interpreter/_import/spec.py:390]:
```suggestion
            # "type" as their __class__.__name__. We check __name__ as well
```

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

**Review by jacobtylerwalls — APPROVED** [review]
Great!

> I think the six test serves as a good test for the general logic. We can now import and find that module where we previously couldn't as we support the _SixMetaPathImporter now.

Indeed. (Sorry, I thought you had said you were about to revert the changes to that test.)

**Comment by DanielNoord** [comment]
> Great!
> 
> > I think the six test serves as a good test for the general logic. We can now import and find that module where we previously couldn't as we support the _SixMetaPathImporter now.
> 
> Indeed. (Sorry, I thought you had said you were about to revert the changes to that test.)

Yeah, that was unclear. My bad!

Thanks for the reviews all. I'm skipping the coverage check on this one!

**Pull request merged**
